src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -11,80 +11,507 @@
+    pipe->my_offset = pipe->my_count * dt_size * rank;
+    if (rank == size - 1) {
+        pipe->my_count += count_total % size;
+    }
+
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+
+    task->allreduce_sliding_window.reduce_in_progress = 0;
+    task->allreduce_sliding_window.barrier_task       = NULL;
+
+    return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+
+out:
+    ucc_tl_ucp_coll_finalize(task->allreduce_sliding_window.allgather_task);
+    tl_error(UCC_TASK_LIB(task), "failed to start allreduce sliding window");
+    return status;
 }
 
 ucc_status_t
-ucc_tl_ucp_allreduce_sliding_window_finalize(ucc_coll_task_t __attribute__((unused)) *coll_task) //NOLINT
+ucc_tl_ucp_allreduce_sliding_window_finalize(ucc_coll_task_t *coll_task)
 {
-    return UCC_OK;
+    ucc_tl_ucp_task_t *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_status_t       st =
+        ucc_ee_executor_finalize(task->allreduce_sliding_window.executor);
+
+    if (ucc_unlikely(st != UCC_OK)) {
+        tl_error(UCC_TASK_LIB(task), "failed to finalize executor");
+    }
+
+    if (ucc_unlikely(ucc_tl_ucp_allreduce_sliding_window_free_gwbi(coll_task)
+            != UCC_OK)) {
+        tl_error(UCC_TASK_LIB(task), "failed to free resources");
+    }
+
+    st = ucc_tl_ucp_coll_finalize(&task->super);
+
+    if (ucc_unlikely(st != UCC_OK)) {
+        tl_error(UCC_TASK_LIB(task), "failed to finalize collective");
+    }
+
+    return st;
 }
 
 static inline void ucc_tl_ucp_allreduce_sliding_window_reduction(
-    ucc_coll_task_t __attribute__((unused)) *coll_task, ucc_tl_ucp_allreduce_sw_buf_t __attribute__((unused)) *accbuf,//NOLINT
-    ucc_tl_ucp_allreduce_sw_buf_t __attribute__((unused)) *getbuf)//NOLINT
+    ucc_coll_task_t *coll_task, ucc_tl_ucp_allreduce_sw_buf_t *accbuf,
+    ucc_tl_ucp_allreduce_sw_buf_t *getbuf)
 {
+    ucc_status_t       status = UCC_OK;
+    ucc_tl_ucp_task_t *task   = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_coll_args_t *  args   = &TASK_ARGS(task);
+    ucc_datatype_t     dt     = TASK_ARGS(task).dst.info.datatype;
+
+    status =
+        ucc_dt_reduce(accbuf->buf, getbuf->buf, accbuf->buf, accbuf->count, dt,
+                      args, 0, 0, task->allreduce_sliding_window.executor,
+                      &task->allreduce_sliding_window.etask);
+
+    task->allreduce_sliding_window.reduce_in_progress = 1;
+
+    if (ucc_unlikely(UCC_OK != status)) {
+        tl_error(UCC_TASK_LIB(task), "failed to perform dt reduction");
+        task->super.status = status;
+        return;
+    }
 }
 
 static inline void
-ucc_tl_ucp_allreduce_sliding_window_test_reduction(ucc_tl_ucp_task_t __attribute__((unused)) *task)//NOLINT
+ucc_tl_ucp_allreduce_sliding_window_test_reduction(ucc_tl_ucp_task_t *task)
 {
+    ucc_status_t status;
+
+    #define SAVE_STATE(_reduce_in_progress) _reduce_in_progress = 1
+
+    EXEC_TASK_TEST(task->allreduce_sliding_window.reduce_in_progress,
```

Comment:
No, it can only happen with reduce_in_progress=1 because this func is called in two places: 1) right after it is set, and 2) within an if (reduce_in_progress) check.

Maybe we can just check whether etask is null, but what would the SAVE_STATE macro be in that case?

Reply:
> Maybe we can just check whether etask is null, but what would the SAVE_STATE macro be in that case?

I fail to see how it is related, since, IIUC, `SAVE_STATE` has no effect (it reassigns `reduce_in_progress` to 1, but it already equaled this value). Checking through `etask==NULL` allows to save one variable, and it is usually how it is done in the codebase if I remember correctly

I am not sure to understand what `SAVE_STATE` is useful for. Could you explain?

Reply:
You're definitely right about `SAVE_STATE` and `etask`. I changed `etask` to `reduce_task` and used that instead of `reduce_in_progress`. Thank you, it's much better this way!